runtime/src/bank/builtin_programs.rs

```diff
@@ -62,3 +62,313 @@ mod tests {
     }
 }
+
+#[cfg(test)]
+mod tests_core_bpf_migration {
+    use {
+        crate::bank::{
+            builtins::{
+                core_bpf_migration::{tests::TestContext, CoreBpfMigrationConfig},
+                BuiltinPrototype, StatelessBuiltinPrototype, BUILTINS, STATELESS_BUILTINS,
+            },
+            tests::{create_genesis_config, create_simple_test_bank},
+            ApplyFeatureActivationsCaller, Bank,
+        },
+        solana_accounts_db::accounts_db::CalcAccountsHashDataSource,
+        solana_sdk::{
+            account::{AccountSharedData, ReadableAccount},
+            clock::DEFAULT_SLOTS_PER_EPOCH,
+            feature::{self, Feature},
+            feature_set::FeatureSet,
+            native_loader,
+            pubkey::Pubkey,
+        },
+        std::sync::Arc,
+        test_case::test_case,
+    };
+
+    enum TestPrototype<'a> {
+        Builtin(&'a BuiltinPrototype),
+        Stateless(&'a StatelessBuiltinPrototype),
+    }
+    impl<'a> TestPrototype<'a> {
+        fn deconstruct(&'a self) -> (&'a Pubkey, &'a CoreBpfMigrationConfig) {
+            match self {
+                Self::Builtin(prototype) => (
+                    &prototype.program_id,
+                    prototype.core_bpf_migration_config.as_ref().unwrap(),
+                ),
+                Self::Stateless(prototype) => (
+                    &prototype.program_id,
+                    prototype.core_bpf_migration_config.as_ref().unwrap(),
+                ),
+            }
+        }
+    }
+
+    #[test_case(TestPrototype::Builtin(&BUILTINS[0]); "system")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[1]); "vote")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[2]); "stake")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[3]); "config")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[4]); "bpf_loader_deprecated")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[5]); "bpf_loader")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[6]); "bpf_loader_upgradeable")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[7]); "compute_budget")]
+    #[test_case(TestPrototype::Builtin(&BUILTINS[8]); "address_lookup_table")]
+    #[test_case(TestPrototype::Stateless(&STATELESS_BUILTINS[0]); "feature_gate")]
+    fn test_core_bpf_migration(prototype: TestPrototype) {
+        let mut bank = create_simple_test_bank(0);
+
+        let (builtin_id, config) = prototype.deconstruct();
+        let feature_id = &config.feature_id;
+        let source_program_id = &config.source_program_id;
+
+        let mut feature_set = FeatureSet::all_enabled();
+        feature_set.inactive.insert(*feature_id);
+        bank.feature_set = Arc::new(feature_set);
+
+        let test_context = TestContext::new(&bank, builtin_id, source_program_id);
+
+        // Simulate crossing an epoch boundary for a new bank.
+        bank.apply_feature_activations(ApplyFeatureActivationsCaller::NewFromParent, false);
```

Comment:
`Bank::apply_feature_activations()` should not be used directly, instead write to the feature account like in this test: https://github.com/anza-xyz/agave/blob/499d36e3549abe12d3a276955aac9d3d354d7d2c/runtime/src/bank/tests.rs#L11878

That way we get the full integration and also the epoch boundary recompilation phase of the program cache which is important as the core programs interact with it.

Reply:
Sounds good I'll rework these tests.

Do the feature activation bits I've added look accurate?

Reply:
Okay @Lichtso I took care of this request!